src/libcharon/plugins/kernel_iph/kernel_iph_net.c

```diff
@@ -16,6 +16,7 @@
 /* Windows 7, for some iphlpapi.h functionality */
 #define _WIN32_WINNT 0x0601
 #include <winsock2.h>
+#include <wincrypt.h>
```

Comment:
Not sure about this (including a header one explicitly does not want to use seems weird). I noticed such a conflict too when building on GitHub (strangely not on AppVeyor), but then added `-DWIN32_LEAN_AND_MEAN` to `CFLAGS` to successfully compile strongSwan without any code changes. I guess we could add that in the configure script when we detect a windows target.

Reply:
> When compiled with `-DNOCRYPT` option, as required to successful compile of `openssl` plugin

I'm not aware that that's necessary. Works perfectly fine without it in all our builds on Windows. Where did you see that?  For example, see the [AppVeyor builds](https://ci.appveyor.com/project/tobiasbrunner/strongswan-52lo9), which run the unit tests with the _openssl_ plugin and that build doesn't require any additional defines at all.

As I mentioned, on GitHub the (experimental) build required `WIN32_LEAN_AND_MEAN` (`NOCRYPT` is not explicitly defined). I don't know exactly what the difference is between GitHub and AppVeyor (we definitely use an older OpenSSL version on the latter) but maybe it's due to a change in the MinGW sources.

Reply:
Well, I'm compiling on Ubuntu 18.04, using packaged mingw-gcc (`i686-w64-mingw32-gcc (GCC) 7.3-win32 20180312`).
I took openssl compiled libraries from [CURL build](https://curl.se/windows/dl-7.74.0_2/openssl-1.1.1i_2-win32-mingw.zip), version 1.1.1i. When compiling `openssl` plugin **without** `-DNOCRYPT` i get tons of compilation errors, starting with:
```
/bin/bash ../../../../libtool  --tag=CC   --mode=compile i686-w64-mingw32-gcc -DHAVE_CONFIG_H -I. -I../../../../../../strongswan/src/libstrongswan/plugins/openssl -I../../../..  -I../../../../../../strongswan/src/libstrongswan -DFIPS_MODE=0 -DDEBUG  -g -O0 -Wall -Wno-pointer-sign -Wno-format-security -I/tmp/openssl-1.1.1i_2-win32-mingw/include -mno-ms-bitfields -include /tmp/build/strongswan-win32-openssl2/config.h -MT openssl_ec_private_key.lo -MD -MP -MF $depbase.Tpo -c -o openssl_ec_private_key.lo ../../../../../../strongswan/src/libstrongswan/plugins/openssl/openssl_ec_private_key.c &&\
mv -f $depbase.Tpo $depbase.Plo

libtool: compile:  i686-w64-mingw32-gcc -DHAVE_CONFIG_H -I. -I../../../../../../strongswan/src/libstrongswan/plugins/openssl -I../../../.. -I../../../../../../strongswan/src/libstrongswan -DFIPS_MODE=0 -DDEBUG -g -O0 -Wall -Wno-pointer-sign -Wno-format-security -I/tmp/openssl-1.1.1i_2-win32-mingw/include -mno-ms-bitfields -include /tmp/build/strongswan-win32-openssl2/config.h -MT openssl_ec_private_key.lo -MD -MP -MF .deps/openssl_ec_private_key.Tpo -c ../../../../../../strongswan/src/libstrongswan/plugins/openssl/openssl_ec_private_key.c  -DDLL_EXPORT -DPIC -o .libs/openssl_ec_private_key.o
In file included from /usr/share/mingw-w64/include/windows.h:95:0,
                 from /usr/share/mingw-w64/include/winsock2.h:23,
                 from ../../../../../../strongswan/src/libstrongswan/utils/compat/windows.h:24,
                 from ../../../../../../strongswan/src/libstrongswan/utils/utils.h:39,
                 from ../../../../../../strongswan/src/libstrongswan/library.h:101,
                 from ../../../../../../strongswan/src/libstrongswan/credentials/builder.h:41,
                 from ../../../../../../strongswan/src/libstrongswan/plugins/openssl/openssl_ec_private_key.h:26,
                 from ../../../../../../strongswan/src/libstrongswan/plugins/openssl/openssl_ec_private_key.c:21:
/tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/x509.h:83:34: error: expected ‘)’ before numeric constant
 typedef STACK_OF(X509_EXTENSION) X509_EXTENSIONS;
                                  ^
/tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/pkcs7.h:47:3: error: expected ‘)’ before numeric constant
 } PKCS7_SIGNER_INFO;
   ^
In file included from /tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/crypto.h:23:0,
                 from /tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/bio.h:20,
                 from /tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/evp.h:16,
                 from ../../../../../../strongswan/src/libstrongswan/plugins/openssl/openssl_ec_private_key.h:24,
                 from ../../../../../../strongswan/src/libstrongswan/plugins/openssl/openssl_ec_private_key.c:21:
/tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/safestack.h:20:32: error: pasting "stack_st_" and "(" does not give a valid preprocessing token
 # define STACK_OF(type) struct stack_st_##type
                                ^
/tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/safestack.h:23:5: note: in expansion of macro ‘STACK_OF’
     STACK_OF(t1); \
     ^~~~~~~~
/tmp/openssl-1.1.1i_2-win32-mingw/include/openssl/safestack.h:130:29: note: in expansion of macro ‘SKM_DEFINE_STACK_OF’
 # define DEFINE_STACK_OF(t) SKM_DEFINE_STACK_OF(t, t, t)
                             ^~~~~~~~~~~~~~~~~~~
...
```

Compilation works fine when CFLAGS contains `-DNOCRYPT` or `-DWIN32_LEAN_AND_MEAN`. The clue about `-DNOCRYPT`` I found somewhere on [strongswan.org list](https://lists.strongswan.org/pipermail//dev/2017-August/001807.html).

Again - if `-DNOCRYPT` nor `-DWIN32_LEAN_AND_MEAN` is not used, the change here is harmless (just `<wincrypt.h>` will be included a little bit earlier), if used - the change is required to succesfully compile `openssl` plugin with my openssl library.


Few more information from Ubuntu's package manager:
```
$ dpkg -l | grep mingw
ii  binutils-mingw-w64        2.30-7ubuntu1+8ubuntu1       all      Cross-binutils for Win32 and Win64 using MinGW-w64
ii  binutils-mingw-w64-i686   2.30-7ubuntu1+8ubuntu1       amd64    Cross-binutils for Win32 (x86) using MinGW-w64
ii  binutils-mingw-w64-x86-64 2.30-7ubuntu1+8ubuntu1       amd64    Cross-binutils for Win64 (x64) using MinGW-w64
ii  g++-mingw-w64             7.3.0-11ubuntu1+20.2build1   all      GNU C++ compiler for MinGW-w64
ii  g++-mingw-w64-i686        7.3.0-11ubuntu1+20.2build1   amd64    GNU C++ compiler for MinGW-w64 targeting Win32
ii  g++-mingw-w64-x86-64      7.3.0-11ubuntu1+20.2build1   amd64    GNU C++ compiler for MinGW-w64 targeting Win64
ii  gcc-mingw-w64             7.3.0-11ubuntu1+20.2build1   all      GNU C compiler for MinGW-w64
ii  gcc-mingw-w64-base        7.3.0-11ubuntu1+20.2build1   amd64    GNU Compiler Collection for MinGW-w64 (base package)
ii  gcc-mingw-w64-i686        7.3.0-11ubuntu1+20.2build1   amd64    GNU C compiler for MinGW-w64 targeting Win32
ii  gcc-mingw-w64-x86-64      7.3.0-11ubuntu1+20.2build1   amd64    GNU C compiler for MinGW-w64 targeting Win64
ii  gdb-mingw-w64             8.0.1-0ubuntu3+10.5          amd64    Cross-debugger for Win32 and Win64 using MinGW-w64
ii  gdb-mingw-w64-target      8.0.1-0ubuntu3+10.5          all      Cross-debugger server for Win32 and Win64 using MinGW-w64
ii  mingw-w64                 5.0.3-1                      all      Development environment targeting 32- and 64-bit Windows
ii  mingw-w64-common          5.0.3-1                      all      Common files for Mingw-w64
ii  mingw-w64-i686-dev        5.0.3-1                      all      Development files for MinGW-w64 targeting Win32
ii  mingw-w64-tools           5.0.3-1                      amd64    Development tools for 32- and 64-bit Windows
ii  mingw-w64-x86-64-dev      5.0.3-1                      all      Development files for MinGW-w64 targeting Win64
ii  win-iconv-mingw-w64-dev   0.0.8-2                      all      iconv implementation using Windows API (for cross-building)
```

Reply:
What is the version of OpensSSL used on AppVeyor builds? Where can I find binaries used on these builds? I could test them with my setup to see where is the problem - openssl version or my minnw setup.

Reply:
I made one more test - with OpenSSL version 1.0.2u used by strongswan CI extracted straight from appveyour's _"Visual Studio 2019"_ machine and this compiles cleanly without `-DNOCRYPT` or `-DWIN32_LEAN_AND_MEAN` flags.

It looks like problems appear with compiling `openssl` plugin using OpenSSL version 1.1.1.

Reply:
Yes, conflicts apparently start with OpenSSL 1.1.0. How about we set `-DNOCRYPT` automatically but only for the _openssl_ plugin when building on Windows? I've pushed a commit to that effect to the branch (also enabled builds against newer OpenSSL versions on AppVeyor and fixed a bug in the plugin).

Reply:
LGTM. But I would rather keep `<wincrypt.h>` inclusion for the case `-DNOCRYPT` would be used for any other reason than openssl plugin.

Reply:
Is there a reason to do so?

Reply:
Well, some third-party plugin requiring `-DWIN32_MEAN_AND_LEAN` or `-DNOCRYPT`?

I mean problem occur with this option on, it is diagnosed, so why just not make it solved? 
Inclusion of `<wincrypt.h>` **is required** for successful compilation of any file which includes `<iprtrmib.h>` (i.e. `kernel_iph_net.c`) , either included implicitly (which is affected by build options) or explicitly which does not harm.

Just my 5¢.

Reply:
As mentioned, my build worked fine with `-DWIN32_MEAN_AND_LEAN`. So I had another look and apparently this problem was already fixed in the MinGW headers four years ago (see [this commit](https://sourceforge.net/p/mingw-w64/mingw-w64/ci/6cc6d2e6e673030cbbc073adae6c28055028effb/)). The fix was released with 6.0.0, which is why the build fails on Ubuntu 18.04 as that only ships MinGW 5.0.3.

Reply:
OK, you're perfectly right.

So everything looks good, compiles cleanly on my Ubuntu18.04/mingw-5.0.3.

Thank you very much for your attention